h2/utilities.py

```diff
@@ -267,9 +243,59 @@ def _reject_pseudo_header_fields(headers, hdr_validation_flags):
         yield header
 
 
+def _validate_host_authority_header(authority_header_val, host_header_val,
+                                    hdr_validation_flags):
+    """
+    Given the :authority and Host headers from a request block that isn't
+    a trailer, check that:
+     1. At least one of these headers is set.
+     2. If both headers are set, they match.
+    Raises a ProtocolError if this is a received header block, or an
+    InvalidHeaderBlockError if this is a header block we are about to send.
+    """
+    # If we have not-None values for these variables, then we know we saw
+    # the corresponding header.
+    authority_present = (authority_header_val is not None)
+    host_present = (host_header_val is not None)
+
+    # If we are client-side, then we are sending the request header block.
+    # If we are server-side, then we are receiving the block.
+    if hdr_validation_flags.is_client:
+        mode = 'sending'
+    else:
+        mode = 'receiving'
+
+    # It is an error for a request header block to contain neither
+    # an :authority header nor a Host header.
+    if not authority_present and not host_present:
+        if mode == 'receiving':
```

Comment:
So I ~think you’re suggesting a function something like:

    def core_validating_logic(authority, host):
        # check at least one is set
        # if both set, check authority == host

which raises a ProtocolError if it fails.

Then the function for sending would catch the `ProtocolError`, an re-raise as an `InvalidHeaderBlockError` with the same exception message. And we tweak the error messages so they don’t refer to receiving/sending, and the same error text can apply to both.

Have I got that right?